### ISSUE TITLE
Assign groups/heats

### ISSUE THREAD

**Issue by jfly** [cubing/ccm#257]
From the WRC:

> The following features would be nice to have in future score/competition organization tools:
* For every round of an event you can define the number of groups you would like to have and the competitors are assigned to their groups automatically.
* For this, I suggest two different approaches:
1. sorting the competitors by their names and dividing this list in the number of groups
2. going over the list and assigning the first competitor to the first group, the second to the second group, etc., and start over when the number of groups is reached.
* Individual changing of assigned groups should be possible as well, maybe also by importing a grouping sheet (.csv file).
* Last but not least the assigned group number should be printed on every individual score sheet.

I personally am pretty opposed to yet another file format, and would rather encode grouping information in an extension to the [WCA Competition JSON Format](https://github.com/cubing/worldcubeassociation.org/wiki/WCA-Competition-JSON-Format).

**Comment by Laura-O** [comment]
I do not really understand how you want to use the JSON format (I guess this is a comment on my suggestion to use a CSV file?).

I think most of the organizers/delegates use spreadsheets to manage the registrations, you can also download a CSV file when you use the WCA website as registration platform. So this is probably the format that is mainly used for handling this data (or you can at least convert it to CSV easily).

**Comment by jfly** [comment]
I think we should be uploading group information to the WCA website. Thus, we need to update the [WCA Competition JSON Format](https://github.com/cubing/worldcubeassociation.org/wiki/WCA-Competition-JSON-Format) (because that's how we submit results). Once we've have a file format that encodes group membership information, we should use it, rather than inventing something new.

**Comment by SAuroux** [comment]
I have a strong feeling that you are talking past each other there. 

From what I understand, @Laura-O ist talking about (maybe even manually created) files for importing groupings into score/competition organization tools (e.g. CCM, cubecomps, etc.). For this, a CSV or XLS file is the obvious choice, because importing registration data into score/competition organization tools currently works the exact same way (e.g. cubecomps, and the registration data from the WCA registration system is exported in CSV format, similar for all other registration system that I know)

Meanwhile @jfly is apparently talking about a file format for importing groupings into the database, together with results. So about a completely different topic, if I understand the situation correctly. Still a short comment to this: As the results team leader, I generally disagree that grouping information should be uploaded together with competition results. Getting accurate grouping information for every competition seems to be a near impossible task, judging from what we can observe from the results team side, and the potential gain of doing that is not quite apparent to me.

**Comment by FatBoyXPC** [comment]
I think it needs to be made clear here that XLS files are much more difficult to work with than CSV. So, if for whatever reason it's agreed that a spreadsheet format is to be used, it should most definitely be CSV.

I do think that JSON for this makes more sense, but I feel like this can be debated by both sides heavily. It will certainly be easier to work with JSON rather than CSV.

@SAuroux, why do you feel that it is an impossible task for groups to be submitted with the results? I feel like competitions with multiple groups/heats have to determine those somewhere, so they know who to call and when. If they have that data, it surely seems trivial at that point to get it into the results (especially if we have the help of some sort of tool on the WCA website).

**Comment by SAuroux** [comment]
I don't question that handling JSON files might be more convenient for a group of programmers. But JSON is defintiely less suitable for random people that might need to create a file manually in addition. ;)

I never said, that it is an impossible task for groups to be submitted with the results. I said, that getting **accurate** grouping information **for every competition** seems to be a near impossible task: 

- Planned groupings almost never perfectly match the scrambles eventually used **for all** competitors:
  * spontaneous group changes
  * heat systems (!!!)
  * etc.
- Only first rounds have planned groupings for obvious reasons
- Not all competition have planned groupings at all
- etc.

Not to mention the huge additional burden for both organisers and results team. I can see a lot of organisers/delegates messing up with this really badly. What is the benefit at all?

TLDR: I don't want inaccurate/incomplete data to be stored in our database, so I prefer to not consider that at all.

**Comment by jfly** [comment]
Those are all good points, @SAuroux. I agree that we should not be uploading incomplete/inaccurate information.

There's a real software problem here, which is that we have 2 data formats, the undocumented and unnamed csv thing, and the WCA Competition JSON format. There are some problems with the CSV format:

1. It's undocumented.
2. From a software communicating with other software perspective, it adds no value, and only increases complexity by making software support 2 formats.
3. It's hard to change. This is partially because of 1 and 2, but is also just the nature of CSV: it's very hard to add a new field without causing a lot of other things to change.

As a thought experiment, could we discuss what would be involved in getting rid of the CSV file format entirely, and moving to JSON? Is there anything besides these 2 points?

- Non coders can't edit it
  - This was raised by @SAuroux. I'd like to understand what sorts of things people are currently doing by hand. Perhaps the only reason people need to open the CSV in excel is because of other software limitations that we could address. @SAuroux, could you provide insight into this?
- There are a number of websites that would have to be changed. This is completely true, but not a reason to not do this. A probably incomplete list:
  - [CubingUSA](http://www.cubingusa.com/) - @jbcm627 @FatBoyXPC 
  - [CubingChina](http://cubingchina.com/) - ([github](https://github.com/CubingChina/cubingchina)) - @Baiqiang
  - [cubecomp.de](https://cubecomp.de/) ([github](https://github.com/fw42/cubecomp)) - @fw42
  - [cuber.com.br](http://www.cuber.com.br/worlds/index.php?location=reg) - @pedrosino
  - [cubecomps](http://cubecomps.com/) ([outdated github](https://github.com/Luis-J-Ianez/cubecomps.com)) - @Luis-J-Ianez

**Comment by SAuroux** [comment]
> I'd like to understand what sorts of things people are currently doing by hand.

I think you are still missing what @Laura-O was aktually talking about: Importing custom groupings into score/competition organization tools. Of course these groupings have to come from somewhere. :)

So far I haven't seen any other tool being used by organizers to make groupings than Excel, so XLS or CSV as export format are the logical consequence.

**Comment by pedrosino** [comment]
I can second what Sebastién said.

I used the csv export from my system both for cubecomps and to make groups,
cards, etc.

If there's a good way to import a .json into excel, I wouldn't mind the
change :)

**Comment by FatBoyXPC** [comment]
I think that's due to not having any sort of provided tool to accomplish this with. This will obviously be solved with CCM (as it will allow for the assignment of groups). This could also easily be created as a simple static page that outputs a JSON file after a web form is filled out. Of course, it could get a lot fancier than a webform where people type names into heats.

While I, too, use the excel export option, that's only because that's currently the easiest way to do things. I actually am not in favor of us doing it this way, or in favor of submitting results **at all**. I much prefer to have the results live on the WCA server, and the "results import" process basically just grabs the results from the respective tables (or collections) OR we just set a flag to mark the results as official. That's another discussion for another thread, though :)

**Comment by jbcm627** [comment]
I imagine a built-in tool for assigning heats would be fine for most use cases. But I can't imagine such a tool would ever be (1) more flexible than Excel and (2) as friendly for non-programmers executing arbitrary macro tasks.  So I think there will always be use cases where Excel is better for assigning heats, necessitating some way of getting heat data to/from Excel.

**Comment by coder13** [comment]
After talking with Jeremy, we determined that what we want to do is start with a dead simple system that ignores staff and simultaneous events and just put people in different heats.

**Comment by jfly** [comment]
Caleb and I agreed that we should extend the WCA Competition JSON to have group information, so people who want to do fancier stuff can write their own scripts to generate whatever groupings they want.

Furthermore, the dead simple group assignment we plan to support in CCM at first will allow you to manually edit it however you want. So for tiny local competitions, you could handle staff by hand.

**Comment by SAuroux** [comment]
On first sight, I don#t understand the connection of "fancy grouping" and
"including the grouping into the competition JSON". So far, we simply do
not care about the Grouping in the results posting process, no matter what
grouping was used.

2016-08-04 0:23 GMT+02:00 Jeremy Fleischman <notifications@github.com>:

**Comment by jfly** [comment]
I'll write up a proper document explaining my vision in September, but the basic idea is to extend the WCA Competition JSON format to be able to encode information like schedule, locations/venues/rooms/stages, timezones, cutoffs, time limits, and heats/groups. Most of this additional information is stuff the WCA Results Team doesn't care about, but I believe that having a standardized format for "all the stuff you need to plan for a competition" would allow all our various competition management software to start talking to each other, and save organizers from having to enter the same information in multiple places. A lot of smart people have made tools to handle each of the individual pieces of organizing a competition, but it's really annoying to compose them, because they don't all speak the same language.

**Comment by Goosly** [comment]
@jfly  I just stumbled upon this thread. Sorry for bumping this after almost 4 years, but look at where we are now, we actually have many tools that talk to each other through the wcif! I want to say thanks to everyone who stubbornly held on to the idea of storing grouping information in a json file - life is so much easier for many organizers and delegates around the world now :-)
